jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -36,23 +43,61 @@
   /** Helper for loading a .class file. */
   private static class ClassFileLoader extends ClassLoader {
 
-    private final Path classFile;
+    private static final Map<Path, Class<?>> cache = new ConcurrentHashMap<>();
```

Comment:
I think we should avoid having to use `ConcurrentHashMap` over `HashMap` unless it's absolutely necessary. Since for most cases, and in our case here, we won't be using this class in concurrent situations, using `ConcurrentHashMap` would slow down operations in the usual cases. I was thinking to just have the cache be an instance field. Do you think we should have a global cache?

Reply:
If it's an instance field we would end up using a new empty cache every time we create a new class loader, which we do a bunch (see L198). I could change it back to a `HashMap` and just leave a comment about it not being threadsafe, considering threading here is still a hypothetical. Or we could make it an instance field of `MainClassFinder`, but then it's still effectively global to `ClassFileLoader`. Or we could change the logic so we only ever instantiate one `ClassFileLoader`, but that's an arbitrary extra constraint, and would require changing some path-finding logic.

Reply:
Hmm, it looks like the line 198 instantiation should be moved out of the loop so there's only one `ClassFileLoader` instantiated.